common/kvstore/tablestore/table_store_builder.go

```diff
@@ -0,0 +1,402 @@
+		// In the future if we change schema versions, we may need to write migration code here.
+		return fmt.Errorf(
+			"incompatible schema version: code is at version %d, data on disk is at version %d",
+			currentSchemaVersion, onDiskSchema)
+	}
+
+	return nil
+}
+
+// This method adds and removes tables as needed to match the given list of tables. The table ID map is updated
+// to reflect the new state of the tables.
+func addAndRemoveTables(
+	base kvstore.Store,
+	metadataTable kvstore.Table,
+	namespaceTable kvstore.Table,
+	tableIDMap map[uint32]string,
+	currentTables []string) error {
+
+	// Determine which tables to keep and which to drop.
+	originalTablesSet := make(map[string]bool)
+	for _, table := range tableIDMap {
+		originalTablesSet[table] = true
+	}
+	newTablesSet := make(map[string]bool)
+	for _, table := range currentTables {
+		newTablesSet[table] = true
+	}
+	tablesToDrop := make([]string, 0, len(originalTablesSet))
+	for table := range originalTablesSet {
+		if !newTablesSet[table] {
+			tablesToDrop = append(tablesToDrop, table)
+		}
+	}
+	tablesToAdd := make([]string, 0, len(newTablesSet))
+	for table := range newTablesSet {
+		if !originalTablesSet[table] {
+			tablesToAdd = append(tablesToAdd, table)
+		}
+	}
+
+	// Drop tables that are not in the list.
+	reverseTableIDMap := make(map[string]uint32)
+	for tableName, tableID := range tableIDMap {
+		reverseTableIDMap[tableID] = tableName
+	}
+	for _, tableName := range tablesToDrop {
+		err := dropTable(base, metadataTable, namespaceTable, tableName, reverseTableIDMap[tableName])
+		if err != nil {
+			return fmt.Errorf("error dropping table %s: %w", tableName, err)
+		}
+		delete(tableIDMap, reverseTableIDMap[tableName])
+	}
+
+	if len(tablesToAdd) == 0 {
+		// No tables to add, bail out early.
+		return nil
+	}
+
+	// Determine the table IDs for the new tables to be added.
+	// We want to fill gaps prior to assigning new IDs.
+	newTableIDs := make([]uint32, 0, len(tablesToAdd))
+	nextID := uint32(0)
+	for i := 0; i < len(tablesToAdd); i++ {
+		for {
+			_, alreadyUsed := tableIDMap[nextID]
+			if !alreadyUsed {
+				newTableIDs = append(newTableIDs, nextID)
+				nextID++
+				break
+			}
+			nextID++
+		}
+	}
+
+	// Sort tables to add. Ensures deterministic table IDs given the same input.
+	sort.Strings(tablesToAdd)
+
+	// Add new tables.
+	for i, tableName := range tablesToAdd {
+		tableID := newTableIDs[i]
+		err := createTable(namespaceTable, tableName, tableID)
+		if err != nil {
+			return fmt.Errorf("error creating table %s: %w", tableName, err)
+		}
+		tableIDMap[tableID] = tableName
+	}
+
+	err := sanityCheckNamespaceTable(namespaceTable, tableIDMap, currentTables)
+	if err != nil {
+		return fmt.Errorf("error sanity checking namespace table: %w", err)
+	}
+
+	return nil
+}
+
+// Perform sanity checks on the namespace table.
+// This method makes potential logic in the namespace errors fail fast and visibly.
+func sanityCheckNamespaceTable(
```

Comment:
Is this for testing/debugging the implementation is correct?

Reply:
This is sanity check code intended to make noise if a bug in the code results in inconsistencies between `tableIDMap`, `namespaceTable`, and the user provided schema. I intended this check to run even in production environments, since if the tables are corrupted we will have really big problems.

That being said, if you don't think it belongs I'm willing to remove it.

Reply:
Sounds fine to keep then